### PR TITLE
feat: add interactive dependency visualization graph

### LINKED ISSUE #352: [Feature]:
### Is your feature request related to a problem?

<html>
<body>
<html><head></head><body><h1>🚀 Feature Request: Interactive Dependency Visualization Graph</h1><h2>📌 Description</h2><p>EnvForge currently provides deterministic ML environment generation and compatibility validation for frameworks like PyTorch, TensorFlow, CUDA, and cuDNN. However, users still need to manually understand how different dependencies interact with each other.</p><p>I would like to propose an <strong>Interactive Dependency Visualization Graph</strong> feature that visually represents the relationship between:</p><ul><li><p>Python versions</p></li><li><p>CUDA versions</p></li><li><p>cuDNN versions</p></li><li><p>GPU drivers</p></li><li><p>ML frameworks (PyTorch / TensorFlow / YOLO)</p></li><li><p>Operating system compatibility</p></li></ul><p>This feature would help users better understand dependency chains, compatibility conflicts, and recommended environment structures through an intuitive graph-based UI.</p><hr><h1>🎯 Objectives</h1><ul><li><p>Improve dependency transparency</p></li><li><p>Help users visualize ML environment relationships</p></li><li><p>Simplify debugging of compatibility issues</p></li><li><p>Provide a modern and interactive developer experience</p></li><li><p>Reduce confusion around CUDA/framework version matching</p></li></ul><hr><h1>✨ Proposed Features</h1><h2>🔹 Interactive Dependency Graph</h2><p>Display dependencies as connected nodes and edges.</p><p>Example relationships:</p><ul><li><p>PyTorch → CUDA</p></li><li><p>CUDA → GPU Driver</p></li><li><p>TensorFlow → Python</p></li><li><p>cuDNN → CUDA</p></li></ul><hr><h2>🔹 Real-Time Compatibility Indicators</h2><p>Each node can display status indicators:</p>
Status | Meaning
-- | --
🟢 Compatible | Stable and supported
🟡 Warning | Minor compatibility concerns
🔴 Conflict | Unsupported combination

<hr><h2>🔹 Hover &amp; Click Details</h2><p>Users can interact with nodes to view:</p><ul><li><p>Supported versions</p></li><li><p>Compatibility notes</p></li><li><p>Warnings</p></li><li><p>Recommendations</p></li><li><p>Installation guidance</p></li></ul><hr><h2>🔹 Graph Filtering Options</h2><p>Allow filtering by:</p><ul><li><p>Framework</p></li><li><p>CUDA version</p></li><li><p>Python version</p></li><li><p>Operating system</p></li><li><p>GPU vendor</p></li></ul><hr><h2>🔹 Suggested Compatibility Paths</h2><p>Highlight recommended dependency chains.</p><p>Example:</p><pre><code class="language-txt">Python 3.10
   ↓
CUDA 12.1
   ↓
PyTorch 2.3
   ↓
cuDNN 9.x
</code></pre><hr><h1>🛠️ Technical Implementation</h1><h2>Frontend (Next.js)</h2><h3>Suggested Libraries</h3><ul><li><p>React Flow</p></li><li><p>D3.js</p></li><li><p>Cytoscape.js</p></li></ul><h3>UI Components</h3><ul><li><p>Interactive graph canvas</p></li><li><p>Dependency cards</p></li><li><p>Status legends</p></li><li><p>Search/filter controls</p></li><li><p>Zoom &amp; pan support</p></li></ul><hr><h2>Backend (FastAPI)</h2><h3>Add:</h3><ul><li><p>Dependency graph generation utility</p></li><li><p>Compatibility relationship endpoint</p></li><li><p>Structured dependency mapping API</p></li></ul><h3>Possible Endpoint</h3><pre><code class="language-http">GET /api/v1/dependencies/graph
</code></pre><h3>Example Response</h3><pre><code class="language-json">{
  "nodes": [
    {
      "id": "cuda-12.1",
      "type": "cuda"
    },
    {
      "id": "pytorch-2.3",
      "type": "framework"
    }
  ],
  "edges": [
    {
      "source": "cuda-12.1",
      "target": "pytorch-2.3"
    }
  ]
}
</code></pre><hr><h1>🔒 Safety Considerations</h1><p>This feature will remain visualization-only and will NOT:</p><ul><li><p>Execute shell commands</p></li><li><p>Modify environments</p></li><li><p>Bypass existing compatibility checks</p></li><li><p>Alter deterministic resolution logic</p></li></ul><p>The graph will only display validated compatibility data already managed by EnvForge.</p><hr><h1>📚 Benefits</h1><ul><li><p>Better developer understanding</p></li><li><p>Easier debugging of dependency conflicts</p></li><li><p>Improved onboarding experience</p></li><li><p>More production-grade UI/UX</p></li><li><p>Enhanced transparency for environment generation</p></li><li><p>Strong educational value for ML beginners</p></li></ul><hr><h1>🧪 Testing Plan</h1><ul><li><p>Test graph rendering with multiple dependency trees</p></li><li><p>Validate node-edge mapping</p></li><li><p>Verify responsiveness across devices</p></li><li><p>Ensure compatibility data accuracy</p></li><li><p>Test large dependency graphs for performance</p></li></ul><hr><h1>📦 Suggested Tech Stack</h1><h2>Frontend</h2><ul><li><p>Next.js</p></li><li><p>TailwindCSS</p></li><li><p>React Flow / D3.js</p></li></ul><h2>Backend</h2><ul><li><p>FastAPI</p></li><li><p>Pydantic</p></li><li><p>Dependency mapping utilities</p></li></ul><hr><h1>🙌 Contribution Plan</h1><p>I would like to work on this feature under NSoC'26.</p><h3>Planned Implementation</h3><ul><li><p>Interactive dependency graph UI</p></li><li><p>Backend dependency mapping endpoint</p></li><li><p>Compatibility status indicators</p></li><li><p>Search &amp; filtering system</p></li><li><p>Responsive graph visualization</p></li><li><p>Testing &amp; documentation</p></li></ul><p>Please assign this issue to me if approved. 🚀</p></body></html>
</body>
</html>

## Describe alternatives you've considered

I considered simpler alternatives such as displaying compatibility relationships in static tables or predefined compatibility matrices. However, those approaches become difficult to understand as dependency chains grow larger and more complex.

Another alternative was using expandable accordion sections for CUDA, Python, and framework compatibility details, but this would still require users to manually trace relationships between dependencies.

An interactive graph-based visualization would provide a much clearer and more intuitive understanding of:

* Dependency relationships
* Compatibility conflicts
* Recommended environment structures
* Framework-to-CUDA mappings

This approach would significantly improve debugging, onboarding, and overall user experience compared to static documentation or table-based layouts.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MistryVishwa**
Screenshot 2026-05-29 162551

## 📌 Overview

This PR introduces an **Interactive Dependency Visualization Graph** to EnvForge, enabling users to visually explore relationships between Python versions, CUDA versions, cuDNN, GPU drivers, and ML frameworks such as PyTorch and TensorFlow.

The feature improves transparency around dependency compatibility and helps users better understand how environment components interact before generating setup scripts.

Closes #352

---

### Interactive Dependency Graph

* Added visual node-based dependency graph
* Displays relationships between:

  * Python
  * CUDA
  * cuDNN
  * GPU Drivers
  * PyTorch
  * TensorFlow
  * Environment Profiles

### Compatibility Status Indicators

* 🟢 Compatible
* 🟡 Warning
* 🔴 Conflict

### Node Interaction

* Clickable dependency nodes
* Detailed compatibility information
* Dependency relationship insights

### Graph Navigation

* Zoom support
* Pan support
* Responsive layout
* Improved visualization experience

### Dependency Mapping

* Dynamic rendering of compatibility relationships
* Structured node-edge architecture
* Clear dependency flow representation

---

### Frontend

* Added Dependency Visualization component
* Implemented interactive graph rendering
* Added status indicators and legends
* Improved responsiveness across devices

### Backend

* Added dependency relationship handling
* Structured compatibility mapping support
* Integrated graph data generation logic

---

## 📚 Benefits

* Better understanding of dependency chains
* Easier identification of compatibility conflicts
* Improved user experience
* Enhanced debugging workflow
* More intuitive environment planning

---

### Functional Testing

* Verified graph rendering
* Verified node interactions
* Verified dependency relationships
* Verified compatibility status indicators

### UI Testing

* Desktop responsiveness
* Tablet responsiveness
* Mobile responsiveness

### Validation Testing

* Tested multiple dependency combinations
* Tested compatibility and conflict scenarios
* Verified graph stability with larger dependency sets

---

## 🙌 Notes

This feature provides a visual and user-friendly way to understand environment dependencies, making EnvForge more accessible for both beginners and experienced ML engineers.

## Summary by CodeRabbit

* **New Features**
  * Added "Dependencies" navigation link to the navbar
  * Introduced interactive Dependency Explorer with two modes: ML Frameworks and System Infrastructure
  * Implemented search filtering and node selection capabilities
  * Added interactive graph visualization with hover connectivity highlighting and panning/zoom controls
  * Included details panel and legend displaying dependency information and node categories

[Review Change Stack](https://app.coderabbit.ai/change-stack/rishabh0510rishabh/EnvForage/pull/377?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by MistryVishwa** [comment]
Hello @rishabh0510rishabh I have completed this project. Please review and merge the PR.

**Review by rishabh0510rishabh — APPROVED** [review]
